### PR TITLE
Link on install modules

### PR DESCRIPTION
**Pull request by roigcarlo**
Expands #5145

It may not be complete but with this changes tests are passing.

### CODE REVIEW

**Comment by roigcarlo** [comment]
@RiccardoRossi I did not use your branch because of the number of changes... just in case there is something wrong

**Comment by loumalouomega** [comment]
@roigcarlo should I update my app?, how do I activate this? (It is not in the description)

**Comment by loumalouomega** [comment]
imagen
This is not good

**Comment by roigcarlo** [comment]
Sorry I should have explained better. CI will not work in this branch unless we change also the scripts to run, as it changes the location of the modules. (Windows only passes because doesn't run the tests)

The aim of the PR was to show the changes needed and test if it was more or less something easy to deploy.

After dealing with the Fluids and the Structural apps I think the easiest way will be to change it by ourselves in all the apps. Those two were battle horses but the rests of the apps looks more or less easy to deal with (Dem will be painful too)

File: applications/FluidDynamicsApplication/python_scripts/adjoint_fluid_analysis.py [hunk 1]
```diff
@@ -3,8 +3,8 @@
 import KratosMultiphysics as Kratos
 import KratosMultiphysics.FluidDynamicsApplication as KFluid
 
-from analysis_stage import AnalysisStage
-from fluid_dynamics_analysis import FluidDynamicsAnalysis
+from KratosMultiphysics.analysis_stage import AnalysisStage
+from KratosMultiphysics.FluidDynamicsApplication.fluid_dynamics_analysis import FluidDynamicsAnalysis
```
Comment (philbucher) [inline applications/FluidDynamicsApplication/python_scripts/adjoint_fluid_analysis.py:7]: I would say that we make separate PRs for this, this will reduce the scope of this PR very much
Also I started doing this / tell ppl to do this for a long time already

New apps already have this, and also many old ones I have ported already (e.g. most of StructuralMechanics and ContactStructuralMechanics FYI @loumalouomega)

this should make your life easier @roigcarlo

Comment (loumalouomega) [reply applications/FluidDynamicsApplication/python_scripts/adjoint_fluid_analysis.py:7]: Can I update my app with this format and it will work with the current system?, OK I will do a PR

Comment (philbucher) [reply applications/FluidDynamicsApplication/python_scripts/adjoint_fluid_analysis.py:7]: yes!
In fact I did most of it already some time ago, e.g. here:
https://github.com/KratosMultiphysics/Kratos/blob/7a12b5ab2f5e3eb8840dafa83ee244273f615429/applications/ContactStructuralMechanicsApplication/python_scripts/contact_structural_mechanics_static_solver.py#L8-L12

**Comment by ddiezrod** [comment]
It might be left for a different PR, but I guess at some point application folders should be removed from path, right?

File: applications/FluidDynamicsApplication/python_scripts/adjoint_fluid_analysis.py (hunk 1 above)
Comment (roigcarlo) [reply applications/FluidDynamicsApplication/python_scripts/adjoint_fluid_analysis.py:7]: :+1: In fact the problem is that half of the apps use the new system and half use the older. Structural was very easy to deal with for example but fluid was a pain.

Also, a little bit out of topic. I see a huge amount of imports in the middle of the code. This causes a lot of errors and it should only be used in case there is no other choice.

**Comment by roigcarlo** [comment]
@ddiezrod Applications folder should not be in the path that I am aware. Only for individual external applications I think, but yes, with the change it won't be necessary anymore

**Comment by philbucher** [comment]
> It might be left for a different PR, but I guess at some point application folders should be removed from path, right?

indeed, you are right @ddiezrod 
This is done already if the [latest version of the app-importing-mechanism](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/python_interface/__init__.py#L66-L78) is used
As you see there nothing is added to sys.path

=> new apps generated with the app-generator use this already, others can be switched once all the python-imports are updated (or an app doesn't have python-scripts)

See the [Wiki](https://github.com/KratosMultiphysics/Kratos/wiki/Applications-as-python-modules)

**Comment by philbucher** [comment]
if you search for `_ImportApplicationAsModule` in the Kratos-Repo you will find all the apps that already use the latest mechanism

**Comment by RiccardoRossi** [comment]
ouch i didn t intend to break anything with this Pr.

i mean proposed changes to the FluidDynamics are for good and should go in, independently on the change of the install mechanism

nevertheless i am afraid we would need some way to gelp the transition.

btw, are we sure this is the way to go before investing more time i to this?

upvote for yes, downvote for no...

**Comment by ddiezrod** [comment]
@RiccardoRossi  I might be wrong but I understand that in the current status of this PR, old imports (without KratosMultiphysics.MyApplication) still work, dont they? @roigcarlo

**Comment by RiccardoRossi** [comment]
as i understand of you do

from KratosMultiphyscs import *

it will work, nevertheless non pythonic imports will not.

for example

import KratosMultiphysics
import analysis_stage

would not (whike it works right now because of. the tricks we do)

Carlos i am correct on this, right?

**Comment by roigcarlo** [comment]
Well it would work but it will create recursive imports that cause that module names appear as not initialized if you use imports in the middle of the code (which we do, a lot).

As @ddiezrod and @philbucher discussed, we can start fixing the scripts before making the change in the module structure, as the new ones will work with the old structure anyway, but we should decide how to tell people.

Also, if you are interested:
`grep -Prn '^\s*(import|from) ((?!(Kratos|sys|os|argv|re|math|json|etc)).)*$'`

That should return the list with most of the invalid imports with the new structure (you can expand the exp to include other keywords like numpy etc...). Beware that the regex itself will not work in the VisualStudioCode project search (the side one) or normal grep (notice the -P flag), but will for a single file in the normal search (the control+F), as the expression requires some advanced regex features.

**Comment by RiccardoRossi** [comment]
well the * import is pythonic...

using it in the middle of the code is forbidden in python3 though...

**Comment by pooyan-dadvand** [comment]
> well the * import is pythonic... using it in the middle of the code is forbidden in python3 though...

It is true. But the problem is when you make from KratosMultiphysics import * in a .py inside the KratosMultiphysics module which creates a recursive import. (similar to when in c++ we include "a.h" inside "b.h" which also includes "a.h")

**Comment by philbucher** [comment]
just adding that we can force the new import system by using the latest version of the app-importer => [`_ImportApplicationAsModule`](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/python_interface/__init__.py#L66) as [described in the Wiki](https://github.com/KratosMultiphysics/Kratos/wiki/Applications-as-python-modules#2-the-second-step-depends-if-a-large-application-which-needs-backwards-compatibility-or-a-small-application-where-backwards-compatibility-is-not-a-problem-eg-because-only-few-people-use-itapplications-that-have-no-python-files)

File: kratos/python_scripts/process_factory.py [hunk 2]
```diff
@@ -28,16 +30,16 @@ def ConstructListOfProcesses( self, process_list ):
 
                 try:
                     full_module_name = kratos_module_name + "." + python_module_name
-                    python_module = __import__(full_module_name,fromlist=[python_module_name])
+                    python_module = importlib.import_module(full_module_name)
                 except ImportError: # old import mechanism for backwards-compatibility
-                    __import__(kratos_module_name)
-                    python_module = __import__(python_module_name)
+                    importlib.import_module(kratos_module_name)
+                    python_module = importlib.import_module(python_module_name)
```
Comment (philbucher) [inline kratos/python_scripts/process_factory.py:36]: why do you change this? 
It took me some time to figure this out. Also I think the behavior of importlib is depending on the version of python (at least some of the functionalities for sure)
I think this is a very sensitive change that should be discussed separately

**Review by philbucher — CHANGES_REQUESTED** [review]
blocking bcs of the change of the importmechanism

File: kratos/python_scripts/process_factory.py (hunk 2 above)
Comment (roigcarlo) [reply kratos/python_scripts/process_factory.py:36]: Didn't we agree on changing this last thursday?
I was under the impresion that we were using __import__ because of importlib being incomatible iwt <=2.6 version of python but concluding that it was fine to change as everybody was using it.

Comment (philbucher) [reply kratos/python_scripts/process_factory.py:36]: ah ok that was that
Why is this better again than `__import__`?

I suggest we make a separate PR then we can update the apps also in the same PR, what do you think?

Comment (roigcarlo) [reply kratos/python_scripts/process_factory.py:36]: Ok to me.

Basically importlib allows cx_freeze, nuitca, etc, detect dynamic imports, while calling __importlib__ does not. I think they hook the call and push the name of the lib in the module path so they know what to pack (or something like that, havn't studied the mechanism in detail)

Comment (pooyan-dadvand) [reply kratos/python_scripts/process_factory.py:36]: I'm not sure about the mechanism but as @roigcarlo said this change is needed for making dynamic imports work with cx_freeze and other packagers. As we have discussed it (and agreed informally) in the committee I asked  @roigcarlo to add it to this PR to have it a complete pack.

Comment (philbucher) [reply kratos/python_scripts/process_factory.py:36]: I did not remember 
Either way, see #5198

Comment (pooyan-dadvand) [reply kratos/python_scripts/process_factory.py:36]: No problem,

As long as there are no conflict between this one and #5198 we are ok with separating it!

Comment (philbucher) [reply kratos/python_scripts/process_factory.py:36]: No conflict, in fact I also ported this to most apps

Comment (pooyan-dadvand) [reply kratos/python_scripts/process_factory.py:36]: I see now. Many thanks!! :+1:

**Comment by ddiezrod** [comment]
Is this ready to be brought back to life after https://github.com/KratosMultiphysics/Kratos/pull/5460 ?

**Comment by RiccardoRossi** [comment]
yes, 
      PR specific changes should be only in CMakeLists.txt

**Comment by philbucher** [comment]
I strongly suggest to open a new PR, this PR changes many files that were ported in the meantime.
As Riccardo says the new PR should only change the CMakeLists

**Comment by pooyan-dadvand** [comment]
The new one is the #5825

**Comment by roigcarlo** [comment]
Closed by #5825
